lms/events/event.py

```diff
@@ -101,3 +103,60 @@ def _get_assignment_id(self):
             return assignment.id
 
         return None
+
+
+@dataclass
+class AuditTrailEvent(BaseEvent):
+    type: EventType.Type = EventType.Type.AUDIT_TRAIL
+
+    instance: BASE = None
+    """Object for which we are tracking changes"""
+
+    action: str = None
+    """What happen to the object: crated,updated,deleted..."""
+
+    source: str = None
+    """In which context the change happen"""
+
+    def _get_data(self):
+        """
+        Fill the event's data value.
+
+        This is called on BaseEvent.__post_init__
+        """
+        changes = {}
+        instance_details = inspect(self.instance)
+        for attr in instance_details.attrs:
+            history = instance_details.get_history(attr.key, True)
+
+            if not history.has_changes():
+                continue
+
+            changes[attr.key] = (
+                history.deleted[0] if history.deleted else None,
+                history.added[0] if history.added else None,
+            )
+
+        return {
+            "model": self.instance.__class__.__name__,
+            "id": self.instance.id,
+            "action": self.action,
+            "source": self.source,
+            # LTI users will have a FK to the User table.
+            # userid is useful for other authentication methods.
+            # For example this will be the user's email while using google oauth.
+            "userid": self.request.identity.userid,
+            "changes": changes,
+        }
+
+    @staticmethod
+    def notify(request: Request, instance: BASE, source="admin_pages"):
```

Comment:
```suggestion
    def notify(request: Request, instance: BASE, action, source):
```

Should we be accepting the `action` here? Then this wouldn't need to change the tests or defaults etc on new things?

Reply:
The way things are strucutred now notify decides the action. The idea is you could for example do:

```
obj = get_or_create_whatever()
...
AuditTrail.notify(obj)
```
And notify would pick the right `action`